libraries/lib-stretching-sequence/AudioSegmentFactory.cpp

```diff
@@ -58,15 +60,22 @@ AudioSegmentFactory::CreateAudioSegmentSequenceForward(double t0) const
       }
       else if (clip->GetPlayEndTime() <= t0)
          continue;
+      const auto propagator = std::make_shared<PitchShiftChangePropagator>();
+      clip->SetPitchShiftChangePublisher(propagator);
+      auto subscriber = std::bind(
+         &PitchShiftChangePropagator::Subscribe, propagator,
+         std::placeholders::_1);
```

Comment:
Or just `std::mem_fn`

Reply:
You may ignore most of a9752ac3a7431db60b6742de5ae9a5b5281ea104 and 250725b5876d32aa78c9114af86f0dff36d0fe24, as I did in a recent commit put in place almost exactly what you described. See, too, that this change is outdated.

Reply:
The constructor of `ClipSegment` isn't passed a function that does subscription.  It is just passed a reference to `Observer::Publisher<PitchRatio>`.  That reference can be stored in `StaffPadTimeAndPitch::Parameters`.  Then `StaffPadTimeAndPitch` in its constructor simply subscribes to the publisher, passing in that lambda, and holding the `Subscription` object which is a sort of smart, weak pointer representing the connection between publisher and subscriber.

Reply:
There are multiple threads involved here, and that means some caution in the use of `Observer`.  Creating and destroying Subscriptions, or destroying the Publisher before the Subscription that it gave out, is safe so long as it is all in the main thread.  Publishing from other threads will be safe so long as the callback does not itself cause Subscriptions to be created or destroyed.  I think that applies.

Reply:
Where is Publish called then?

Reply:
I looked through this confusing stuff again, and now I see, ultimately it's in `WaveClip::SetSemitoneShift`

So then, that can simply call the inherited `Publish`

Reply:
> Creating and destroying Subscriptions, or destroying the Publisher before the Subscription that it gave out, is safe so long as it is all in the main thread. Publishing from other threads will be safe so long as the callback does not itself cause Subscriptions to be created or destroyed.

Incidentally, here it's the other way round: subscription creation and destruction is done on the audio thread ; publication on the main thread.

Just took a look at some of the Observer implementation, couldn't see thread unsafety (but I haven't strained myself too much).
Also tried a worst-case scenario, with very fast subscription creation and destructions, concurrently with very fast publications. Couldn't get the program to behave incorrectly, which is more evidence that I can take your word for it.

Reply:
That “inverse” pattern should be safe too.

Is there no way to lift the making and destroying of subscriptions to the main thread?

Shared and weak pointers manage the shared state in a thread safe way — but provided each thread uses different pointer objects.

It is possible to have races in the use of the identical pointer object in two threads.  Don’t do that.

Observer uses weak pointers.

Reply:
The thread safety issue should be commented!